src/mlpack/methods/random_forest/bootstrap.hpp

```diff
@@ -18,30 +18,210 @@ namespace mlpack {
+ */
+class IdentityBootstrap
+{
+ public:
+  template<bool UseWeights,
+           typename MatType,
+           typename LabelsType,
+           typename WeightsType>
+  void Bootstrap(const MatType& dataset,
+                 const LabelsType& labels,
+                 const WeightsType& weights,
+                 MatType& bootstrapDataset,
+                 LabelsType& bootstrapLabels,
+                 WeightsType& bootstrapWeights)
+  {
+    MakeAlias(bootstrapDataset, dataset, dataset.n_rows, dataset.n_cols);
+    MakeAlias(bootstrapLabels, labels, labels.n_elem);
+    if (UseWeights)
+      MakeAlias(bootstrapWeights, weights, weights.n_elem);
+  }
+};
+
+/**
+ * Sequential bootstrap.
+ *
+ * @tparam IndMatType Indicator matrix type.
+ */
+template<typename IndMatType = arma::umat>
+class SequentialBootstrap
+{
+ public:
+  /**
+   * Constructor.
+   *
+   * @param[in] intervals See ComputeAverageUniqueness().
+   * @param[in] colIndicesCount Number of data points in the dataset.
+   */
+  SequentialBootstrap(const IndMatType& intervals,
+    arma::uword colIndicesCount) :
+    intervals(intervals),
+    sampleCount(colIndicesCount),
+    colIndices(arma::linspace<arma::uvec>(0,
+      colIndicesCount - 1,
+      colIndicesCount))
+  {
+    if (intervals.n_cols != 2)
+      throw std::invalid_argument(
+        "SequentialBootstrap::SequentialBootstrap(): "
+        "intervals must be a 2xm matrix!");
+  }
+
+  template<bool UseWeights,
+    typename MatType,
+    typename LabelsType,
+    typename WeightsType>
+  void Bootstrap(const MatType& dataset,
+    const LabelsType& labels,
+    const WeightsType& weights,
+    MatType& bootstrapDataset,
+    LabelsType& bootstrapLabels,
+    WeightsType& bootstrapWeights)
+  {
+    if (labels.n_cols != intervals.n_rows)
+      throw std::invalid_argument("SequentialBootstrap::Bootstrap(): "
+        "labels n_cols and intervals n_rows differ!");
+
+    // observations are stored as columns and dimensions
+    // (number of features) as rows.
+    if (colIndices.n_rows != dataset.n_cols) {
+      throw std::invalid_argument("SequentialBootstrap::Bootstrap(): "
+        "constructed with colIndicesCount different from "
+        "dataset.n_cols!");
+    }
+
+    const arma::uvec phi(ComputeSamples());
+
+    bootstrapDataset = dataset.cols(phi);
+    bootstrapLabels = labels.cols(phi);
+    if (UseWeights)
+      bootstrapWeights = weights.rows(phi);
+  }
+
+  /**
+   * Compute the samples of the next draw.
+   *
+   * @return A list of indices referring to the observations that should
+   *         be sampled.
+   */
+  arma::uvec ComputeSamples() const
+  {
+    std::vector<arma::u64> phi;
+
+    phi.reserve(sampleCount);
+    while (phi.size() < sampleCount)
+    {
+      const arma::vec prob(ComputeNextDrawProbabilities(phi,
+        intervals, colIndices.n_elem));
+      DiscreteDistribution d(
+        DiscreteDistribution(std::vector<arma::vec>(1u, prob)));
```

Comment:
To avoid more allocations than necessary, I would construct `d` once outside the loop, and then set `d.Probabilities()` inside the loop:

```
d.Probabilities() = ComputeNextDrawProbabilities(...);
```